docs/src/rfcs/002-shader-stage-interpolants.md

```diff
@@ -0,0 +1,188 @@
+
+### Decorators
+
+Similar to DirectX, Vulkan has the following decorations that can be applied to interpolants.
+
+| Vulkan | DirectX equivalent
+| ---- | ---- |
+| - | linear |
+| Flat| nointerpolation
+| NoPerspective| noperspective
+| Centroid | centroid
+| Sample| sample
+
+Vulkan has some additional explanations about these decorations as well:
+
+> A variable decorated with `Flat` will not be interpolated. Instead, it will have the same value for every fragment within a triangle. This value will come from a single provoking vertex. A variable decorated with `Flat` can also be decorated with Centroid or Sample, which will mean the same thing as decorating it only as `Flat`.
+
+## Metal shading language
+
+__Mostly todo.__
+
+One thing to note as a different to SPIR-V, GLSL and HLSL is that MSL doesn't seem to require explicit binding semantics or `Location` specifiers. It's specification has a `[[attribute(N)]]` specifier, but it's only used on tesselation control points and vertex shader inputs. It's specifically not used for communication between shader stages.
+
+# SPIR-V Practicalities
+
+Vertex shaders allow `Location` and `Component` decorations on input variable declarations. The `Location` decoration specifies which vertex input attribute is used to read and interpret the data that a variable will consume. The `Component` decoration allows the location to be more finely specified for scalars and vectors, down to the individual components within a location that are consumed. The components within a location are 0, 1, 2, and 3. A variable starting at component N will consume components N, N+1, N+2, …​ up through its size. For single precision types, it is invalid if the sequence of components gets larger than 3.
+
+
+## Proposal
+
+```rust
+// these `attribute`'s would match the input assembler setup on the CPU
+struct VertexInput {
+    #[attribute(0)]
+    position: Vec3,
+    #[attribute(1)]
+    normal: Vec3,
+    #[attribute(2)]
+    color: Vec3,
+}
+
+// naming this is hard, since vertex shader won't really know what shader
+// stages will run after it until the pipeline state object gets created
+// in vulkan. a tesselation shader may run after this for example,
+// and that decision won't be known by our compiler.
+struct VertexOutput {
+    #[position]
+    position: Vec3,
+    normal: Vec3,
+    color: Vec3,
+}
+
+fn vs_main(#[stage_in] input: &VertexInput) -> VertexOutput {
+    VertexOutput {
+        position: world_to_clip.mul(input.position),
+        normal: input.normal,
+        color: input.color,
+    }
+}
+
+struct PixelOutput {
+    // albedo and material id packed into same RGBA8 render target
+    #[render_target(0)]
+    albedo: Vec3,
+    #[render_target(0)]
+    material_idx: u8,
+
+    // normal and exponent packed into RGBA8 render target
```

Comment:
Does SPIR-V support doing this packing like this built-in, or is this something we need to implement ourselves? If ourselves, what's the justification? Seems like a really obscure feature.

Reply:
Vec4 -> RGBA8 already happens automatically of course. And it's relatively common to be (bit) packing data into render-targets (example: gbuffer / vbuffer laydown). However, judging by the sketchy support (shader-playground)[http://shader-playground.timjones.io/2a474cd68f38f6891cadeebe41f394b3] we probably should just only support f32 (and vectors thereof) types in fragment shader outputs.

Reply:
I'm confused, the shader-playground link doesn't have the same packing thing here, and so I'm not sure if you're saying that SPIR-V supports doing this packing built-in, or if we have to implement it ourselves. To be clear, I'm talking about having >1 field mapping to the same render target, and concatenating those fields into a single output.

Reply:
> To be clear, I'm talking about having >1 field mapping to the same render target, and concatenating those fields into a single output.

Ah - I thought you meant the part about packing stuff together with u8 types and such. From the looks of it, it looks like the `Component` decoration isn't valid for render-target outputs and so all packing should be done manually.

In an earlier draft of this I actually had the pixel shader return a fixed-length array of (float|vec) from the pixel shader, something along these lines, which I feel is quite a bit simpler, but might run into trouble when outputting to render-targets with different components, or when outputting depth for example. 

```rust
fn ps_main(#[stage_in] input: &VertexOutput, shading: &ShadingInputs, indirect_lighting: &IndirectLighting) -> [Vec4; 2] {
    let mut T = brdf(&shading);
    T += input.color;
    T += gi(&indirect_lighting);

    [
        pack_gbuffer0(T, material_idx),
        pack_gbuffer1(normal, specular_exponent),
    ]
}
```

I think maybe the best option could be to have the ps output a struct where each member represents one render target. What do you think?